src/core/symbology/qgssymbollayerreference.h

```diff
@@ -0,0 +1,119 @@
+/***************************************************************************
+ qgssymbollayerreference.h
+ ---------------------
+ begin                : June 2019
+ copyright            : (C) 2019 by Hugo Mercier / Oslandia
+ email                : infos at oslandia dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSSYMBOLLAYERREFERENCE_H
+#define QGSSYMBOLLAYERREFERENCE_H
+
+#include "qgis_sip.h"
+#include "qgis_core.h"
+#include <QList>
+#include <QVariant>
+
+class QgsVectorLayer;
+
+/**
+ * We may need stable references to symbol layers, when pointers to symbol layers is not usable
+ * (when a symbol or a feature renderer is cloned for example).
+ * This class provides a way for a QgsFeatureRenderer to express a unique and stable identifier
+ * for one of its symbol layers.
+ *
+ * Such an identifier must be convertible to an integer and constructible from an integer.
+ * It must also have a validity state. A default constructed reference is invalid.
+ * Each feature renderer is free to use the kind of reference it will.
+ */
+class CORE_EXPORT QgsSymbolLayerId
```

Comment:
@troopa81 It is more a QPair<QString, int> where the string is the identifier of a symbol and the int an index of a symbol layer inside this symbol.
This might be reduced to an index after all, but with this "pair", it seems to me a bit more "stable", especially for rule based renderers: a global index would be changed anytime a new rule is added or a symbol of another rule changed. With a symbol key stored, each symbol has its own symbol layer index.

Reply:
ok, thanks for the explanation